controllers/virtualmachineclass/virtualmachineclass_controller.go

```diff
@@ -111,5 +121,176 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+// given VirtualMachineClass. The name of the instance object contains
+// the a hash of SHA256 digest of the ConfigSpec (and other relevant
+// fields) to uniquely identify a configuration. The instance also
+// contains:
+//   - An annotation containing the same hash
+//   - A label with the corresponding VM class name to allow for
+//     efficient server side queries to list instances for a given VM
+//     class.
+//   - A label to indicate if an instance is active. An instance is
+//     marked inactive if the configuration of the VM class is modified,
+//     or the class is removed from the namespace or deleted.  There
+//     can only be one active instance of a VM class. So, all other
+//     instances are marked inactive.
+//   - Owner reference to the owning class.
+func (r *Reconciler) reconcileInstance(ctx *pkgctx.VirtualMachineClassContext) error {
+	hash, err := vmClassHash(ctx, ctx.VMClass)
+	if err != nil {
+		ctx.Logger.Error(err, "Failed to calculate the hash value required to create VM class instance")
+		return err
+	}
+
+	vmClassInstance := &vmopv1.VirtualMachineClassInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ctx.VMClass.Namespace,
+			Name:      ctx.VMClass.Name + "-" + hash,
+		},
+	}
+
+	// Note that there's a window where users will be able to specify
+	// a stale VM class instance to create a VM if the call comes in
+	// after we create a new instance but before we mark the existing
+	// active one as "inactive". The alternative is to flip the order
+	// but that will lead to rejection of VM creations even though
+	// there's a valid class associated with the namespace. So we
+	// choose former.
+	_, err = ctrlutil.CreateOrPatch(ctx, r.Client, vmClassInstance, func() error {
+		if err := ctrlutil.SetControllerReference(
+			ctx.VMClass, vmClassInstance, r.Scheme()); err != nil {
+
+			return err
+		}
+
+		vmClassInstance.Spec.VirtualMachineClassSpec = ctx.VMClass.Spec
+
+		vmClassInstance.Annotations = map[string]string{
+			pkgconst.VirtualMachineClassHashAnnotationKey: hash,
+		}
+
+		vmClassInstance.Labels = map[string]string{
+			vmopv1.VMClassInstanceActiveLabelKey: "",
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	var list vmopv1.VirtualMachineClassInstanceList
+
+	if err := r.Client.List(ctx, &list,
+		client.InNamespace(ctx.VMClass.Namespace)); err != nil {
+
+		ctx.Logger.Error(err, "Error listing instances for VM class")
+		return err
+	}
+
+	for _, instance := range list.Items {
+		// Only process instances owned by this VM class
+		owned, err := ctrlutil.HasOwnerReference(instance.OwnerReferences, ctx.VMClass, r.Scheme())
+		if err != nil {
+			return fmt.Errorf("failed to check if VM class instance is owned by the class. err: %w", err)
+		}
+
+		if !owned {
+			ctx.Logger.Info("Skipping this instance since it is not owned by VM class")
+			continue
+		}
+
+		// Create patch helper before modifying the instance
+		patchHelper, err := patch.NewHelper(&instance, r.Client)
```

Comment:
I don't think you need all of what the patcher helper does here - just use Client.Patch instead